tag_tracer.go

```diff
@@ -0,0 +1,265 @@
+	if err != nil {
+		log.Warnf("unable to create decaying delivery tag: %s", err)
+		return
+	}
+	t.decaying[topic] = tag
+}
+
+func (t *tagTracer) removeDeliveryTag(topic string) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.decaying, topic)
+}
+
+func (t *tagTracer) decayingDeliveryTag(topic string) (connmgr.DecayingTag, error) {
+	if t.decayer == nil {
+		return nil, fmt.Errorf("connection manager does not support decaying tags")
+	}
+	name := fmt.Sprintf("pubsub-deliveries:%s", topic)
+
+	return t.decayer.RegisterDecayingTag(
+		name,
+		GossipSubConnTagDecayInterval,
+		connmgr.DecayFixed(GossipSubConnTagDecayAmount),
+		connmgr.BumpSumBounded(0, GossipSubConnTagMessageDeliveryCap))
+}
+
+func (t *tagTracer) bumpDeliveryTag(p peer.ID, topic string) error {
+	t.RLock()
+	defer t.RUnlock()
+
+	tag, ok := t.decaying[topic]
+	if !ok {
+		return fmt.Errorf("no decaying tag registered for topic %s", topic)
+	}
+	return tag.Bump(p, GossipSubConnTagBumpMessageDelivery)
+}
+
+func (t *tagTracer) bumpTagsForMessage(p peer.ID, msg *Message) {
+	for _, topic := range msg.TopicIDs {
+		err := t.bumpDeliveryTag(p, topic)
+		if err != nil {
+			log.Warnf("error bumping delivery tag: %s", err)
+		}
+	}
+}
+
+// nearFirstPeers returns the peers who delivered the message while it was still validating
+func (t *tagTracer) nearFirstPeers(msg *Message) []peer.ID {
+	t.Lock()
+	defer t.Unlock()
+	peersMap, ok := t.nearFirst[t.msgID(msg.Message)]
+	if !ok {
+		return []peer.ID{}
+	}
+	peers := make([]peer.ID, 0, len(peersMap))
+	for p := range peersMap {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+// -- internalTracer interface methods
+var _ internalTracer = (*tagTracer)(nil)
+
+func (t *tagTracer) AddPeer(p peer.ID, proto protocol.ID) {
+	t.tagPeerIfDirect(p)
+}
+
+func (t *tagTracer) Join(topic string) {
+	t.addDeliveryTag(topic)
+}
+
+func (t *tagTracer) DeliverMessage(msg *Message) {
+	nearFirst := t.nearFirstPeers(msg)
+
+	t.bumpTagsForMessage(msg.ReceivedFrom, msg)
+	for _, p := range nearFirst {
+		t.bumpTagsForMessage(p, msg)
+	}
+
+	// delete the delivery state for this message
+	t.Lock()
+	delete(t.nearFirst, t.msgID(msg.Message))
+	t.Unlock()
+}
+
+func (t *tagTracer) Leave(topic string) {
+	t.removeDeliveryTag(topic)
+}
+
+func (t *tagTracer) Graft(p peer.ID, topic string) {
+	t.tagMeshPeer(p, topic)
+}
+
+func (t *tagTracer) Prune(p peer.ID, topic string) {
+	t.untagMeshPeer(p, topic)
+}
+
+func (t *tagTracer) ValidateMessage(msg *Message) {
+	t.Lock()
+	defer t.Unlock()
+
+	// create map to start tracking the peers who deliver while we're validating
+	id := t.msgID(msg.Message)
+	if _, exists := t.nearFirst[id]; exists {
+		return
+	}
+	t.nearFirst[id] = make(map[peer.ID]struct{})
+}
+
+func (t *tagTracer) DuplicateMessage(msg *Message) {
+	t.Lock()
+	defer t.Unlock()
+
+	id := t.msgID(msg.Message)
+	peers, ok := t.nearFirst[id]
+	if !ok {
+		// the peers map should have been created in ValidateMessage
```

Comment:
that comment is not technically accurate and is confusing; the duplicate might arrive after we have Delivered/Rejected the message.